libmamba/src/core/satisfiability_error.cpp

```diff
@@ -0,0 +1,284 @@
+            }
+
+
+        private:
+            MSolver const& m_solver;
+            MPool const& m_pool;
+            graph_t m_graph;
+            conflict_map m_conflicts;
+            std::map<SolvId, node_id> m_solv2node;
+            node_id m_root_node;
+
+
+            /**
+             * Add a node if it is not already present and return the node id.
+             */
+            auto ensure_solvable(SolvId solv_id, node_t&& pkg_info) -> node_id;
+            auto ensure_solvable(SolvId solv_id, PackageInfo&& pkg_info) -> node_id;
+            auto ensure_solvable(SolvId solv_id, std::string&& dep) -> node_id;
+
+            auto add_conflict(node_id n1, node_id n2) -> void;
+            auto add_expanded_deps_edges(node_id from_id, SolvId dep_id, edge_t const& edge)
+                -> void;
+            auto parse_problems() -> void;
+        };
+
+        auto ProblemsGraphCreator::ensure_solvable(SolvId solv_id, node_t&& node) -> node_id
+        {
+            if (auto const iter = m_solv2node.find(solv_id); iter != m_solv2node.end())
+            {
+                return iter->second;
+            }
+            auto const node_id = m_graph.add_node(std::move(node));
+            m_solv2node[solv_id] = node_id;
+            return node_id;
+        };
+
+        auto ProblemsGraphCreator::ensure_solvable(SolvId solv_id, PackageInfo&& pkg_info)
+            -> node_id
+        {
+            return ensure_solvable(solv_id, PackageNode{ std::move(pkg_info) });
+        }
+
+        auto ProblemsGraphCreator::ensure_solvable(SolvId solv_id, std::string&& dep) -> node_id
+        {
+            return ensure_solvable(solv_id, UnresolvedDependencyNode{ std::move(dep) });
+        }
+
+        auto ProblemsGraphCreator::add_conflict(node_id n1, node_id n2) -> void
+        {
+            m_conflicts[n1].insert(n2);
+            m_conflicts[n2].insert(n2);
+        }
+
+        auto ProblemsGraphCreator::add_expanded_deps_edges(node_id from_id,
+                                                           SolvId dep_id,
+                                                           edge_t const& edge) -> void
+        {
+            for (const auto& solv_id : m_pool.select_solvables(dep_id))
+            {
+                auto pkg_info = PackageInfo(pool_id2solvable(m_pool, solv_id));
+                auto to_id = ensure_solvable(solv_id, std::move(pkg_info));
+                m_graph.add_edge(from_id, to_id, edge);
+            }
+        }
+
+        auto ProblemsGraphCreator::parse_problems() -> void
+        {
+            for (const auto& problem : m_solver.all_problems_structured())
+            {
+                // These functions do not return a reference so we make sure to
+                // compute the value only once.
+                // TODO change name of these functions to make explicit it is not a ref.
+                auto source = problem.source();
```

Comment:
Good catch! `problem.source()` is an `std::optional` so if we can pinpoint a bug we should fix it in `SolverProblem`.

Reply:
@syslaila if you have a reproducible example, I'm willing to investigate. Best would be if you can reproduce using `libmamba`/`libmambapy` from `main` and open a separate issue for it.

Reply:
I'll get an example today. However, I think we shouldn't block the merging of this PR on this since we don't support yet an "UPDATE" solver rule (and I think we would also need to filter out the ids from https://github.com/syslaila/mamba-fork/pull/3/files#diff-e214b9fbd2b0497c09f4b5738a8aec7bdb8f107f5c7e3415be65d12898fc91f3R267)